### ISSUE TITLE
Remote Code Execution Vulnerability in dependency of NServiceBus.AzureFunctions.Worker.ServiceBus

### ISSUE THREAD

**Issue by sanfordn** [Particular/NServiceBus.AzureFunctions.Worker.ServiceBus#449]
### Symptoms

A vulnerable release of Azure.Identity is included in the `.azurefunctions` subdirectory of the build output.

### Who's affected

Anyone using the NServiceBus.AzureFunctions.Worker.ServiceBus package without explicitly updating transitive dependencies is affected.

### Root cause

The Azure Functions SDK uses a novel project generation method during a build that utilizes dependencies not detectable by normal NuGet vulnerability tooling. Details can be found in [this GitHub comment](https://github.com/Particular/NServiceBus.AzureFunctions.Worker.ServiceBus/issues/449#issuecomment-1932791881).

## Fixes

* https://github.com/Particular/NServiceBus.AzureFunctions.Worker.ServiceBus/pull/454 on `master`
* https://github.com/Particular/NServiceBus.AzureFunctions.Worker.ServiceBus/pull/455 on release branch `release-4.2`
* https://github.com/Particular/NServiceBus.AzureFunctions.Worker.ServiceBus/pull/456 on release branch `release-3.1`

#### Description
We use veracode to scan our code, and there was a vulnerability announced for Azure.Identity 1.6.0. See [CVE-2023-36414](https://nvd.nist.gov/vuln/detail/CVE-2023-36414) for more details on that issue. Every time we build our Azure Function, it pulls in a reference to Azure.Identity 1.6.0. We cannot get this dependency out of our code without switching to an entirely different type of Azure Function.

We have traced this back to the .azurefunctions folder. You can clearly see in the function.deps.json that it is referencing Azure.Identity 1.6.0. I'm not entirely sure where this is all pulled in from because we don't seem to have a ton of control of what packages are pulled into the .azurefunctions folder when compiling. 

The function.deps.json looks something like this:
```json
{
  "runtimeTarget": {
    "name": ".NETCoreApp,Version=v6.0",
    "signature": ""
  },
  "compilationOptions": {},
  "targets": {
    ".NETCoreApp,Version=v6.0": {
      "Microsoft.Azure.Functions.Worker.Extensions/1.0.0": {
        "dependencies": {
          "Microsoft.Azure.WebJobs.Extensions.ServiceBus": "5.4.0", // VERSION WITH VULNERABILITY
          "Microsoft.NET.Sdk.Functions": "4.2.0",
          "Microsoft.NETCore.Targets": "3.0.0"
        },
...

      "Microsoft.Azure.WebJobs.Extensions.ServiceBus/5.4.0": {
        "dependencies": {
          "Azure.Messaging.ServiceBus": "7.8.0",
          "Microsoft.Azure.WebJobs": "3.0.33",
          "Microsoft.Extensions.Azure": "1.2.0" // VERSION WITH VULNERABILITY
        },
...
      "Microsoft.Extensions.Azure/1.2.0": {
        "dependencies": {
          "Azure.Core": "1.24.0",
          "Azure.Identity": "1.6.0", // VERSION WITH VULNERABILITY
...
  }
}
```
So the vulnerability is introduced to my application referencing `NServiceBus.AzureFunctions.Worker.ServiceBus 4.2.2` which references `Microsoft.Azure.WebJobs.Extensions.ServiceBus 5.4.0` which references `Microsoft.Extensions.Azure 1.2.0` which references `Azure.Identity 1.6.0`

We also previously referenced `NServiceBus.AzureFunctions.Worker.ServiceBus 4.0.0-rc.2` and this vulnerability exists there too, and seemingly all versions after 3.1.0.

I can provide the full function.deps.json if that would help this situation.
#### Expected behavior
`NServiceBus.AzureFunctions.Worker.ServiceBus 4.2.2+` references `Microsoft.Azure.WebJobs.Extensions.ServiceBus >= 5.13.5` which references an updated version of Azure.Identity and removes the vulnerability that exists in `Azure.Identity 1.6.0`.
#### Actual behavior
`NServiceBus.AzureFunctions.Worker.ServiceBus 4.2.2` references `Microsoft.Azure.WebJobs.Extensions.ServiceBus 5.4.0` and introduces a vulnerability due to its dependency of `Azure.Identity 1.6.0`.
#### Versions
`NServiceBus.AzureFunctions.Worker.ServiceBus 3.1.0`
`NServiceBus.AzureFunctions.Worker.ServiceBus 3.1.1`
`NServiceBus.AzureFunctions.Worker.ServiceBus 4.0.0-beta.2`
`NServiceBus.AzureFunctions.Worker.ServiceBus 4.0.0-rc.2`
`NServiceBus.AzureFunctions.Worker.ServiceBus 4.0.0`
`NServiceBus.AzureFunctions.Worker.ServiceBus 4.1.0`
`NServiceBus.AzureFunctions.Worker.ServiceBus 4.1.1`
`NServiceBus.AzureFunctions.Worker.ServiceBus 4.2.0`
`NServiceBus.AzureFunctions.Worker.ServiceBus 4.2.2`

### Steps to reproduce

- Create Azure Function with .net 6 or .net 8
- Reference `NServiceBus.AzureFunctions.Worker.ServiceBus 3.1.0` or greater
- Build the application and see the function.deps.json in the .azurefunctions folder

#### Workarounds
N/A unless we can force a package version

#### Possible solutions
`NServiceBus.AzureFunctions.Worker.ServiceBus 4.2.2+` references `Microsoft.Azure.WebJobs.Extensions.ServiceBus >= 5.13.5` which references an updated version of Azure.Identity and removes the vulnerability that exists in `Azure.Identity 1.6.0`.

**Comment by ramonsmits** [comment]
You should be able to add an explicit reference to `Azure.Identity` and use a non-vulnerable version of that package. Meanwhile I'm going to check our builds.

**Comment by DavidBoike** [comment]
I've been looking into this, and something doesn't add up here.

I looked at the [dependencies of NServiceBus.AzureFunctions.Worker.ServiceBus 4.2.2](https://www.nuget.org/packages/NServiceBus.AzureFunctions.Worker.ServiceBus/4.2.2) which includes `Microsoft.Azure.Functions.Worker.Extensions.ServiceBus (>= 5.4.0 && < 6.0.0)` and even the [dependencies for the lowest version 5.4.0](https://www.nuget.org/packages/Microsoft.Azure.Functions.Worker.Extensions.ServiceBus/5.4.0#dependencies-body-tab) only includes `Microsoft.Azure.Functions.Worker.Extensions.Abstractions (>= 1.1.0)` which doesn't have any dependencies itself.

We have an internal app that uses our own Azure Functions isolated worker package on .NET 8 and it doesn't bring in Azure.Identity 1.6.0. It has version 1.10.3, including in the `.azurefunctions` directly.

I created a new Functions app with the Isolated Worker model on .NET 8, added the NServiceBus functions package, and it doesn't pull in Azure.Identity at all.

Then I pulled down the [sample Ramon mentioned](https://docs.particular.net/samples/azure-functions/service-bus-worker/) for .NET 7. After building it, the `Project\bin\Debug\net7.0` doesn't contain an Azure.Identity.dll file. `Project\obj\project.assets.json` doesn't mention Azure.Identity either. However `Project\bin\Debug\net7.0\.azurefunctions\` _DOES_ contain Azure.Identity.dll version 1.6.0.

I think your problem is actually with the Functions SDK that gets used for .NET versions before net8.0 and is ultimately responsible for building that `.azurefunctions` folder to do whatever gRPC magic it does in the isolated worker hosting model.

In summary, I'm not sure there's anything that Particular Software can do to fix this for you.

**Comment by DavidBoike** [comment]
Ok, we dug further. Big thanks to @bording for his forensic MSBuild spelunking talents.

Functions is doing some funny stuff here. There's an MSBuild target that actually constructs a new csproj file in a temp folder, dynamically fills it based on the triggers you have in your function app, builds it, and then copies the output to that `.azurefunctions` folder.

Each Extensions project has a sort of "buddy package" that is identified (by name and version) through an assembly-level attribute. So if you're using `Microsoft.Azure.Functions.Worker.Extensions.ServiceBus` package and you have a ServiceBus trigger, it looks for the assembly attributes on the Extensions.ServiceBus assembly, and is directed to a specific version of `Microsoft.Azure.WebJobs.Extensions.ServiceBus` (notice WebJobs in the middle) and _that_ is the thing that has the dependency on Azure.Identity.

So in effect, Microsoft has created a sort of hidden transitive dependency you can't view in NuGet. And if you have a dependency on the Extensions package but no actual ServiceBus trigger, then you don't actually get Azure.Identity showing up in the `.azurefunctions` directory at all. That's why I wasn't seeing it when setting up a brand new Functions project.

This is still solvable on the client side (immediately) by taking an explicit dependency on `Microsoft.Azure.WebJobs.Extensions.ServiceBus` version either 5.13.4 or 5.13.5 (all older ones lead toward a vulnerable Azure.Identity) rather than relying on the transitive dependency in `NServiceBus.AzureFunctions.Worker.ServiceBus`.

If you don't, our package relies on `Microsoft.Azure.Functions.Worker.Extensions.ServiceBus` 5.4.0 or greater (but < 6) which will resolve to the minimum 5.4.0. That has no direct or transitive dependency on Azure.Identity (so it won't show up in the bin directory) but if you download the package, extract it, and examine the DLL in dotPeek, it has the assembly attribute:

```csharp
[assembly: ExtensionInformation("Microsoft.Azure.WebJobs.Extensions.ServiceBus", "5.4.0")]
```

And if you look up [those dependencies](https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.ServiceBus/5.4.0#dependencies-body-tab) you find [Microsoft.Extensions.Azure >= 1.2.0](https://www.nuget.org/packages/Microsoft.Extensions.Azure/1.2.0), which then has the dependency on Azure.Identity >= 1.6.0.

Long story short, you can fix this immediately by updating Microsoft.Azure.WebJobs.Extensions.ServiceBus to latest, but we will be releasing a patch that raises our minimum version of that to current.

**Comment by DavidBoike** [comment]
Updates have been released. [Here's the announcement](https://discuss.particular.net/t/nservicebus-azurefunctions-worker-servicebus-4-2-3-and-3-1-2-patch-releases-available/3834).
